File: project_2/container.cpp
//------------------------------------------------------------------------------
// container.cpp - содержит функции обработки контейнера.
//------------------------------------------------------------------------------

#include "container.h"
#include "stdio.h"

//------------------------------------------------------------------------------
// Конструктор контейнера.
container::container() {
    animals = new animal*[max_len];
    for (int i = 0; i < max_len; ++i) {
        animals[i] = NULL;
    }
    len = 0;
}

// Ввод содержимого контейнера из указанного файла.
int container::In(FILE *input) {
    int amount;
    int input_amount = fscanf(input,"%ld\n", &amount);

    if (input_amount != 1) {
        printf("Wrong format!\n");
        return 1;
    }

    if (amount < 0 || amount > max_len){
        printf("Amount of items in container must be < %ld and > 0.\n", sizeof(container));
        return 1;
    }

    for(int i = 0; i < amount; ++i){
        animal* anim = animal::In(input);
        if (anim == NULL){
            printf("Couldn't create item with index = %d.\n", i);
            return 1;
        }
        if (len >= max_len) {
            printf("Container is full.\n");
            return 1;
        }
        animals[len] = anim;
        ++len;
    }
    return 0;
}

// Случайный ввод содержимого контейнера.
int container::InRnd(int size, FILE *input_g) {
    if (size <0 || size > max_len) {
        printf("Container size must be > 0 and < %d.\n", max_len);
        return 1;
    } else {
        fprintf(input_g,"%d\n",size);
    }
    for (int i = 0; i < size; ++i) {
        animal* anim = animal::InRnd();
        animals[len] = anim;
        ++len;
        anim->OutGenerated(input_g);
    }
    return 0;
}

// Вывод содержимого контейнера в указанный поток.
void container::Out(FILE *output) {
    fprintf(output,"Container contains %d elements.\n", len);
    for (int i = 0; i < len; ++i) {
        fprintf(output,"%d : ",i);
        animals[i]->Out(output);
    }
}

// Прямая сортировка выбором по убыванию.
void container::StraightSelection() {
        for (int i = 0; i < len-1; ++i) {
        animal* current = animals[i];
        int max = i;
        for(int j = i + 1; j < len; ++j) {
            if(animals[j]->CountQuotient() > animals[i]->CountQuotient()){
                animals[i] = animals[j];
                max = j;
            }
        }
        animals[max] = current;
    }
}

// Деструктор контейнера.
container::~container() {
    for (int i = 0; i < len; ++i) {
        delete animals[i];
    }
    len = 0;
    delete animals;
}

File: project_3/container.py
from animals import Animal

MAX_LENGTH = 10000


class Container:
    """
    Класс, описывающий контейнер.
    """
    animals: list

    def __init__(self):
        """
        Конструктор объекта.
        """
        self.animals = list()

    def __str__(self):
        """
        Строковое представление информации об объекте.
        :return: строка с информацией.
        """
        return f'Container contains {len(self.animals)} elements.\n '

    @staticmethod
    def in_from_file(input_file):
        """
        Статический метод ввода информации об объекте из файла.
        :param input_file: входной файл.
        :return: объект, созданный по указанным в файле параметрам.
        """
        container = Container()
        input_str = input_file.readline()
        try:
            amount = int(input_str)
            if amount < 0 or amount > 10000:
                print(f'Amount of items in container must be < {MAX_LENGTH} and > 0.\n')
                return None
            for i in range(amount):
                if not container.add(Animal.in_from_file(input_file)):
                    break
            return container
        except ValueError:
            print('Couldn\'t read container from file!\n')
            return None

    def add(self, animal) -> bool:
        """
        Метод добавления животного в контейнер.
        :param animal: добавляемое животное.
        :return: истина, если добавление прошло успешно, иначе ложь.
        """
        if len(self.animals) < MAX_LENGTH:
            self.animals.append(animal)
            return True
        else:
            print('Container is full.\n')
            return False

    @staticmethod
    def in_rnd(amount):
        """
        Статический метод генерации рандомного контейнера.
        :param amount: размер контейнера.
        :return: в случае корректности заданного размера возвращается сгенерированный контейнер, иначе None.
        """
        container = Container()
        if amount < 0 or amount > 10000:
            print(f'Amount of items in container must be < {MAX_LENGTH} and > 0.\n')
            return None
        for i in range(amount):
            container.add(Animal.in_rnd())
        return container

    def out(self, output_file):
        """
        Вывод информации о контейнере в файл.
        :param output_file: выходной файл.
        :return: -
        """
        output_file.write(str(self))
        for i in range(0, len(self.animals)):
            output_file.write(f'{i} : ' + str(self.animals[i]))

    def out_generated(self, input_g_file):
        """
        Вывод сгенерированных параметров для создания контейнера в файл.
        :param input_g_file: входной файл.
        :return: -
        """
        input_g_file.write(str(len(self.animals)) + "\n")
        for animal in self.animals:
            animal.out_generated(input_g_file)

    def straight_selection(self):
        """
        Метод сортировки выбором объектов по заданному параметру сравнения (по возрастанию).
        :return: -
        """
        for i in range(0, len(self.animals) - 1):
            current = self.animals[i]
            max_index = i
            for j in range(i + 1, len(self.animals)):
                if self.animals[j].count_quotient() > self.animals[i].count_quotient():
                    self.animals[i] = self.animals[j]
                    max_index = j
            self.animals[max_index] = current
